### PR TITLE
Cleanup exported kokkos options

### PR DESCRIPTION
**Pull request by dalg24**
Honor `Kokkos_OPTIONS_NOT_TO_EXPORT` list (that is used when building as part of Trilinos) and add a few options that it makes no sense to export to the list.
For instance, no one care whether I enabled self containment tests when finding an install of Kokkos.

We might want to clean further but I think it is already a step in the right direction.

Here are the changes on my local `KokkosConfigCommon.cmake`
```diff
--- develop
+++ this_pr
@@ -1,5 +1,5 @@
 SET(Kokkos_DEVICES THREADS)
-SET(Kokkos_OPTIONS DEPRECATION_WARNINGS;TESTS;COMPILER_WARNINGS;LAUNCH_COMPILER;CUDA_LAMBDA;COMPLEX_ALIGN)
+SET(Kokkos_OPTIONS DEPRECATION_WARNINGS;LAUNCH_COMPILER;CUDA_LAMBDA;COMPLEX_ALIGN)
 SET(Kokkos_TPLS LIBDL;THREADS)
 SET(Kokkos_ARCH )
 SET(Kokkos_CXX_COMPILER "/Library/Developer/CommandLineTools/usr/bin/c++")

### CODE REVIEW

**Comment by dalg24** [comment]
To elaborate a bit, I am working towards properly exporting variables such as `Kokkos_ENABLE_{CUDA,HIP}_RELOCATABLE_DEVICE_CODE` which is needed to enable building Trilinos against an external install of Kokkos.  And my angle is to do something along these lines
```diff
diff --git a/cmake/KokkosConfigCommon.cmake.in b/cmake/KokkosConfigCommon.cmake.in
index 63949e561..597aff5b0 100644
--- a/cmake/KokkosConfigCommon.cmake.in
+++ b/cmake/KokkosConfigCommon.cmake.in
@@ -10,6 +10,9 @@ SET(Kokkos_CXX_STANDARD @KOKKOS_CXX_STANDARD@)
 FOREACH(DEV ${Kokkos_DEVICES})
   SET(Kokkos_ENABLE_${DEV} ON)
 ENDFOREACH()
+FOREACH(OPT ${Kokkos_OPTIONS})
+  SET(Kokkos_ENABLE_${OPT} ON)
+ENDFOREACH()

 IF(Kokkos_ENABLE_CUDA)
   SET(Kokkos_CUDA_ARCHITECTURES @KOKKOS_CUDA_ARCHITECTURES@)
```
I just don't want us to export all sort of garbage variables.

**Comment by masterleinad** [comment]
I'm wondering if we should rather whitelist options to export.

**Comment by dalg24** [comment]
Both have pros and cons.  If no one feel too strongly about it I would just go with the blacklist.  We might accidentally export more than we intended but we will fix it.

**Comment by masterleinad** [comment]
I would prefer if we already do the
```
FOREACH(OPT ${Kokkos_OPTIONS})
+  SET(Kokkos_ENABLE_${OPT} ON)
+ENDFOREACH()
```
in this pull request so that we have more uniform behavior already between building as part of `Trilinos` or stand-alone.

File: cmake/KokkosConfigCommon.cmake.in [hunk 1]
```diff
@@ -6,10 +6,18 @@ SET(Kokkos_CXX_COMPILER "@CMAKE_CXX_COMPILER@")
...
-# These are needed by KokkosKernels
+# Export Kokkos_ENABLE_<BACKEND> for each backend that was enabled.
+# NOTE: "Devices" is a little bit of a misnomer here.  These are really
+# backends, e.g. Kokkos_ENABLE_OPENMP, Kokkos_ENABLE_CUDA, Kokkos_ENABLE_HIP,
+# or Kokkos_ENABLE_SYCL.
 FOREACH(DEV ${Kokkos_DEVICES})
   SET(Kokkos_ENABLE_${DEV} ON)
 ENDFOREACH()
+# Export relevant Kokkos_ENABLE<OPTION> variables, e.g.
+# Kokkos_ENABLE_CUDA_RELOCATABLE_CODE, Kokkos_ENABLE_DEBUG, etc.
```
Comment (dalg24) [inline cmake/KokkosConfigCommon.cmake.in:17]: uggg is messed that up it is missing the "device".  will wait to see if more changes are requested before I fix

**Comment by dalg24** [comment]
Ping

File: cmake/kokkos_functions.cmake [hunk 2]
```diff
@@ -6,7 +6,12 @@
 # upper-case version for use within
 
 set(Kokkos_OPTIONS_NOT_TO_EXPORT
-  Kokkos_ENABLE_TESTS  Kokkos_ENABLE_EXAMPLES)
+  Kokkos_ENABLE_BENCHMARKS
+  Kokkos_ENABLE_EXAMPLES
+  Kokkos_ENABLE_TESTS
+  Kokkos_ENABLE_HEADER_SELF_CONTAINMENT_TESTS
+  Kokkos_ENABLE_COMPILER_WARNINGS
+)
```
Comment (masterleinad) [inline cmake/kokkos_functions.cmake:14]: I would probably add `DEPRECATION_WARNINGS` if you want to `Kokkos_ENABLE_CUDA_RELOCATABLE_CODE` anyway.

Comment (dalg24) [reply cmake/kokkos_functions.cmake:14]: The deprecation warnings are different.  Although I am not convinced it is really useful, in contrast to the ones I listed, it is something that is observable downstream.

**Review by masterleinad — APPROVED** [review]
I think exporting the options is the important change here. We can also work on the list of non-exported options elsewhere.

File: cmake/kokkos_functions.cmake (hunk 2 above)
Comment (masterleinad) [reply cmake/kokkos_functions.cmake:14]: Sure. I can see that `KOKKOS_ENABLE_DEPRECATED_CODE_*` might be relevant but `DEPRECATION_WARNINGS` could also be an impl variable.

File: cmake/KokkosConfigCommon.cmake.in (hunk 1 above)
Comment (dalg24) [reply cmake/KokkosConfigCommon.cmake.in:17]: Typo corrected.  Amended the change and push forced.

**Comment by cz4rs** [comment]
> Both have pros and cons. If no one feel too strongly about it I would just go with the blacklist. We might accidentally export more than we intended but we will fix it.

One argument for whitelist would be readability: it's probably more important to see which options we allow to be exported versus things we don't.

**Review by cz4rs — APPROVED** [review]
Looks good to me, no strong preference on whitelist vs blacklist.

**Pull request merged**